core/templates/dev/head/domain/topic_viewer/read-only-topic-object.factory.ts

```diff
@@ -0,0 +1,155 @@
+
+/**
+ * @fileoverview Factory for creating instances of ReadOnlyTopic from
+ * topic data domain objects.
+ */
+
+import { downgradeInjectable } from '@angular/upgrade/static';
+import { Injectable } from '@angular/core';
+
+import { SkillSummary, SkillSummaryObjectFactory } from
+  'domain/skill/SkillSummaryObjectFactory';
+import { StoryReference, StoryReferenceObjectFactory } from
+  'domain/topic/StoryReferenceObjectFactory';
+import { StorySummary } from
+  'domain/story/StorySummaryObjectFactory';
+import { Subtopic, SubtopicObjectFactory } from
+  'domain/topic/SubtopicObjectFactory';
+
+export interface ISkillDescriptions {
+  [skillId: string]: string | null;
+}
+
+export interface IDegreesOfMastery {
+  [skillId: string]: number | null;
+}
+
+export class ReadOnlyTopic {
+  _topicName: string;
+  _topicId: string;
+  _canonicalStories: Array<StorySummary>;
+  _additionalStories: Array<StorySummary>;
+  _uncategorizedSkills: Array<SkillSummary>;
+  _subtopics: Array<Subtopic>;
+  _degreesOfMastery: IDegreesOfMastery;
+  _skillDescriptions: ISkillDescriptions;
+
+  constructor(topicName: string, topicId: string,
+      canonicalStories: Array<StorySummary>,
+      additionalStories: Array<StorySummary>,
+      uncategorizedSkills: Array<SkillSummary>, subtopics: Array<Subtopic>,
+      degreesOfMastery: IDegreesOfMastery,
+      skillDescriptions: ISkillDescriptions) {
+    this._topicName = topicName;
+    this._topicId = topicId;
+    this._canonicalStories = canonicalStories;
+    this._additionalStories = additionalStories;
+    this._uncategorizedSkills = uncategorizedSkills;
+    this._subtopics = subtopics;
+    this._degreesOfMastery = degreesOfMastery;
+    this._skillDescriptions = skillDescriptions;
+  }
+
+  getTopicName(): string {
+    return this._topicName;
+  }
+
+  getTopicId(): string {
+    return this._topicId;
+  }
+
+  getCanonicalStories(): Array<StorySummary> {
+    return this._canonicalStories.slice();
+  }
+
+  getAdditionalStories(): Array<StorySummary> {
+    return this._additionalStories.slice();
+  }
+
+  getUncategorizedSkills(): Array<SkillSummary> {
+    return this._uncategorizedSkills.slice();
+  }
+
+  getSubtopics(): Array<Subtopic> {
+    return this._subtopics.slice();
+  }
+
+  getDegreesOfMastery(): IDegreesOfMastery {
+    return this._degreesOfMastery;
+  }
+
+  getSkillDescriptions(): ISkillDescriptions {
+    return this._skillDescriptions;
+  }
+}
+
+@Injectable({
+  providedIn: 'root'
+})
+
+export class ReadOnlyTopicObjectFactory {
+  _subtopicObjectFactory: SubtopicObjectFactory;
+  _skillSummaryObjectFactory: SkillSummaryObjectFactory;
+
+  constructor(
+    private subtopicObjectFactory: SubtopicObjectFactory,
+    private skillSummaryObjectFactory: SkillSummaryObjectFactory) {
+    this._subtopicObjectFactory = subtopicObjectFactory;
+    this._skillSummaryObjectFactory = skillSummaryObjectFactory;
+  }
+
+  getStorySummaryArray(storyDicts: any): Array<StorySummary> {
```

Comment:
Is this really any?

Reply:
Actually, it had one of the fields missing in dict. So, I didn't find relevant exact type.